### ISSUE TITLE
QuestPDF.Drawing.Exceptions.DocumentDrawingException: 'Could not find an appropriate font fallback for the following glyphs:  $U-000D '

### ISSUE THREAD

**Issue by rverberne-carapax** [QuestPDF/QuestPDF#838]
Hi,

I am trying version 2024.3.0.

But what does this error mean?

`
QuestPDF.Drawing.Exceptions.DocumentDrawingException: 'Could not find an appropriate font fallback for the following glyphs:  $U-000D '
`

Is this documented? If so, i have not been able to find it.

**Comment by danirzv** [comment]
FontFallback is the font which get used when QuestPdf could not find specified character in your default font, it's my sample code i use and works fine
```
        if (!string.IsNullOrEmpty(_options.FontPath))
        {
            FontManager.RegisterFont(File.OpenRead(_options.FontPath));
        }

        if (!string.IsNullOrEmpty(_options.FallBackFontPath))
        {
            FontManager.RegisterFont(File.OpenRead(_options.FallBackFontPath));
        }
Document.Create(pdf =>
        {
            pdf.Page(pageHandler =>
            {
                pageHandler.If(
                    !string.IsNullOrEmpty(_options.FontName),
                    x => x.DefaultTextStyle(text => 
                        text.FontFamily(_options.FontName)
                        .Fallback(text => text.FontFamily(_options.FallBackFontName)).DirectionAuto()));
```

**Comment by rverberne-carapax** [comment]
@danirzv Thanks.

The exception message also includes the following text which i did not see (The VS exception popup was sized so that it did not show all the text):

```
Possible solutions: 
1) Install fonts that contain missing glyphs in your runtime environment. 
2) Configure the fallback TextStyle using the 'TextStyle.FontFamilyFallback' method. 
3) Register additional application specific fonts using the 'FontManager.RegisterFont' method. 

You can disable this check by setting the 'Settings.CheckIfAllTextGlyphsAreAvailable' option to 'false'. 
However, this may result with text glyphs being incorrectly rendered without any warning.'
```

**Comment by BC89** [comment]
Seeing the same and I find it perplexing since I'm not specifying a particular font family anywhere and would expect the quest to use it's default. I'm running .net8 in a Linux container. I've tried to specify a few common fonts that I believe exists on both Win and Linux but still no luck.

**Comment by rverberne-carapax** [comment]
I added below line. But i dont know if the is 'the way to go'. I dont like hacky workarounds, but what i dont like more are exceptions that kill my app... And there is no visual difference in my case anyway.

```csharp
QuestPDF.Settings.CheckIfAllTextGlyphsAreAvailable = false;
```

**Comment by eros404** [comment]
I also encountered this problem and I'm not specifying any particular font family.

For me this is due to a call to the `Line()` function on a `TextDescriptor`:

```csharp
 private void ComposeHeader(TableCellDescriptor header)
{
    header
        .Cell()
        .AlignRight()
        .Text(text =>
        {
            text.Line("actuelle").Bold(); // <= Here
            text.Span("Date").Bold();
        });
}
```

When I remove this line, it works fine.
So I'm guessing that the `Line()` function add `\r\n` to the string and that these are not handled by the default font.

**Comment by MarcinZiabek** [comment]
Thank you all for reporting this issue and sharing your feedback 😄 

> The `Line()` function works as expected when I set `QuestPDF.Settings.CheckIfAllTextGlyphsAreAvailable = false`.

@eros404  Thank you for this observation. It looks like the `\r` symbol (added only on the Windows environment) is not supported by many fonts and/or Skia. I will change the implementation so that only a new line character (`\n`) is added. 

The fix will be released soon.

**Comment by henkmollema** [comment]
Hardcoding a `\n` as newline characters works as well (as workaround). E.g.:
```cs
text.Span("\nHello World");
```

**Comment by Krm1t** [comment]
@MarcinZiabek I hope it's okay that i'm adding to this issue as i believe it is related.

This is reproduceable in version 2024.3.1

Just to give you an idea of our setup. Users are writing multiline text in a windows application. The text is then stored in a database. A docker container then generates a pdf. So it's a mix of windows and linux platforms.

Because input happens in windows, new line is added as `/r/n`.
As per this issue, `/r` isn't supported, or more precisely `/r/n` isn't supported because what i found is that if you do `/r /n` then it does actually seem to work.

I'm no expert on the subject but if i were to guess i would think that `/r/n` somehow gets interpreted as some special char rather than 2 individual chars.

I've added a small project that can show this. Initially it will fail but if you go to MyDocument and add a space between /r/n then it will run just fine.
While writing this i decided to try and run the project on windows (IIS Express) and i get the same results.

[QuestPdfNewLine.zip](https://github.com/QuestPDF/QuestPDF/files/15075082/QuestPdfNewLine.zip)

**Comment by merinofg** [comment]
Can also confirm issue with version `2024.3.1`

Running / debugging in my M1 without any issue. All renders properly by default.
As soon as I put it in a docker container, I get the glyphs error message mentioned in this thread.

Forcing to load a font via `FontManager.RegisterFont(File.OpenRead("wwwroot/fonts/Roboto-Regular.ttf"));` only works locally.

My `Dockerfile` in case it helps to reproduce the issue:
```dockerfile
FROM mcr.microsoft.com/dotnet/aspnet:8.0 AS base
USER $APP_UID
WORKDIR /app

FROM mcr.microsoft.com/dotnet/sdk:8.0 AS build
ARG BUILD_CONFIGURATION=Release
WORKDIR /src
COPY ["WebApp/WebApp.csproj", "WebApp/"]
RUN dotnet restore "WebApp/WebApp.csproj"
COPY . .
WORKDIR "/src/WebApp"
RUN dotnet build "WebApp.csproj" -c $BUILD_CONFIGURATION -o /app/build

FROM build AS publish
ARG BUILD_CONFIGURATION=Release
RUN dotnet publish "WebApp.csproj" -c $BUILD_CONFIGURATION -o /app/publish /p:UseAppHost=false

FROM base AS final
WORKDIR /app
COPY --from=publish /app/publish .
ENTRYPOINT ["dotnet", "WebApp.dll"]
```

**Comment by MarcinZiabek** [comment]
Thank you for the collaboration on fixing this issue 😄 

I have applied one more enhancement. Since many font files do not contain the `carriage return` glyph, the mentioned exception could be thrown even for valid environments. I implemented one more step that removes the `\r` character from text content, hopefully completely fixing this problem.

@Krm1t Your findings have been critical. Thank you for investigating it so profoundly!

Would you please test the newest 2024.3.2 version? https://github.com/QuestPDF/QuestPDF/releases/tag/2024.3.2

**Comment by BC89** [comment]
I just updated to the latest stable - 2024.3.2 and can confirm this is still an open issue. Same error as before running on a windows box with docker enabled and docker file identical to merinofg.

**Comment by MarcinZiabek** [comment]
> I just updated to the latest stable - 2024.3.2 and can confirm this is still an open issue. Same error as before running on a windows box with docker enabled and docker file identical to merinofg.

Would you provide an example, where you think the `\r` character is placed and throws an exception? Is it maybe used within the block similar to this one?

```csharp
text.CurrentPageNumber().Format(x => /* formating code that injects the new line with \r */);
```

**Comment by Krm1t** [comment]
> Thank you for the collaboration on fixing this issue 😄
> 
> I have applied one more enhancement. Since many font files do not contain the `carriage return` glyph, the mentioned exception could be thrown even for valid environments. I implemented one more step that removes the `\r` character from text content, hopefully completely fixing this problem.
> 
> @Krm1t Your findings have been critical. Thank you for investigating it so profoundly!
> 
> Would you please test the newest 2024.3.2 version? https://github.com/QuestPDF/QuestPDF/releases/tag/2024.3.2

@MarcinZiabek I'm happy it helped.
I'm sorry it took me so long to get back to you but i ran into some unrelated issues which i'll be creating a new issue about unless someone already did.

That being said i can confirm that the solution seems to have helped, both in our actual application and in the test application that i attached in my earlier comment.

Thanks for the quick fix.

**Comment by MarcinZiabek** [comment]
@merinofg Indeed, it is an entirely separate problem. There is a great chance that the 2024.3.3 release fixes it. Would you please give it a try?

**Comment by merinofg** [comment]
Can confirm that version `2024.3.3` works as intended for me.
Thanks a lot @MarcinZiabek for such fast responses and help! 🙌🏼

**Comment by BC89** [comment]
Latest 2024.3.3 solved my issue referenced above. Thanks!

**Comment by MarcinZiabek** [comment]
The problem appears to have already been fixed.

Since there has been no recent activity, I’m closing this issue. If you have a similar request in the future, please feel free to open a new GitHub issue. Thanks for your feedback! ❤️
